File: combination-sum/combination-sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> result= new ArrayList<>();
        helper(target, 0, candidates, new LinkedList<Integer>(), result);
        return result;
        
    }
    public void helper(int target, int first, int[] candidates, LinkedList<Integer> currList,List<List<Integer>> result){
        int sum = 0;
        for(int i : currList){
            sum+=i;
        }
        
        if(sum == target){
            result.add(new ArrayList(currList));
            return;
        }
        else if(sum>target){
            return;
        }
        else{
            
            while(first<candidates.length){
                
                currList.add(candidates[first]);
                helper(target, first, candidates, currList,result);
                currList.removeLast();
                first++;
            }
            
            
        }
        
    }

File: combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        
        result = []
        
        
        def dfs(idx, currList, currSum):
            
            if currSum == 0:
                result.append(currList[:])
                return 
            if currSum < 0:
                return
            
            
            for i in range(idx, len(candidates)):
                
                currList.append(candidates[i])
                currSum -= candidates[i]
                dfs(i, currList, currSum)
                currList.pop()
                currSum+= candidates[i]
        
        dfs(0, [], target)
        
        return result
            
                
            